README.md
# arb-convert

![Build Status](https://github.com/trufi-association/arb-convert/workflows/build/badge.svg)
[![NPM version](https://img.shields.io/npm/v/arb-convert.svg?style=flat)](https://www.npmjs.com/package/arb-convert)
[![Coverage Status](https://coveralls.io/repos/github/trufi-association/arb-convert/badge.svg?branch=master)](https://coveralls.io/github/trufi-association/arb-convert?branch=master)
[![Package dependencies](https://img.shields.io/david/trufi-association/arb-convert.svg)](https://david-dm.org/trufi-association/arb-convert)
[![GitHub license](https://img.shields.io/github/license/trufi-association/arb-convert.svg)](https://github.com/trufi-association/arb-convert/blob/master/LICENSE)

Tool to convert JSON-based [Application Resource Bundle](https://github.com/google/app-resource-bundle) (ARB) localization format by Google to another format and back.

```shell
npm install -g arb-convert
```

## Background

ARB is the format of choice when using the [Dart intl](https://pub.dev/packages/intl) package to localize [Flutter](https://flutter.dev/) apps. It has not seen wide adoption by translation providers however. Since December 2019, Google shut down its Translator Toolkit that has natively supported the ARB format. So now Flutter devs can either migrate to another library and format, edit files by hand or convert to another format. This project aims to aid with the latter.

## Supported formats

* XLIFF 1.2/2.0/2.1
* GNU gettext
* ... add your format, PRs welcome

Please note that ICU placeholders and plural notations are not converted.

## Node.js

```js
const arbConvert = require('arb-convert');
const { convertFromArb, parseToArb } = arbConvert;

const xmlString = convertFromArb('xliff-1.2', {
    source: '{ "@@locale": "en-US", ... }',
    target: '{ "@@locale": "de-DE", ... }', // optional
    original: 'namespace', // optional
    sourceLanguage: 'en-US', // optional (override)
    targetLanguage: 'de-DE', // optional (override)
});

const result = parseToArb('xliff', {
    content: xmlString,
});
// { source, target, original, sourceLanguage, targetLanguage }

```

## Binaries

### arb2xliff

```shell
Usage: arb2xliff [options]

Options:
  --sourcefile <filename>  source ARB file (required)
  --targetfile <filename>  target ARB file
  --original <value>       where the translations come from
  --sourcelang <locale>    source locale override, e.g. en-US, in case it cannot be determined from file content or file name
  --targetlang <locale>    target locale override, e.g. de-DE, in case it cannot be determined from file content or file name
  --outversion <version>   XLIFF version, e.g. 1.2 (default) or 2.0/2.1
  --out <filename>         write XLIFF to file if given or stdout if omitted
  -h, --help               output usage information
```

### xliff2arb

```shell
Usage: xliff2arb [options]

Options:
  --file <filename>       source XLIFF file (required)
  --sourceout <filename>  write source ARB to file if given or stdout if omitted
  --targetout <filename>  write target ARB to file if given
  -h, --help              output usage information
```

### arb2po

```shell
Usage: arb2po [options]

Options:
  --sourcefile <filename>  source ARB file (required)
  --targetfile <filename>  target ARB file
  --original <value>       where the translations come from
  --sourcelang <locale>    source locale override, e.g. en-US, in case it cannot be determined from file content or file name
  --targetlang <locale>    target locale override, e.g. de-DE, in case it cannot be determined from file content or file name
  --out <filename>         write PO to file if given or stdout if omitted
  -h, --help               output usage information
```

### po2arb

```shell
Usage: po2arb [options]

Options:
  --file <filename>       source PO file (required)
  --sourceout <filename>  write source ARB to file if given or stdout if omitted
  --targetout <filename>  write target ARB to file if given
  -h, --help              output usage information
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.1.2] - 2020-04-21

### Fixed

- Fix error on key mismatch in source and target input files in arb convert [#2](https://github.com/trufi-association/arb-convert/issues/2)

### Security

- Update dependencies and fix audit issues

## [1.1.1] - 2020-01-02

### Fixed

- Add arb2po and po2arb as npm bins

## [1.1.0] - 2020-01-02

### Added

- GNU gettext PO format support
- Full test coverage

## [1.0.1] - 2019-12-30

### Fixed

- Removed unnecessary files from NPM release

## [1.0.0] - 2019-12-29

### Added

- XLIFF 1.2/2.0/2.1 format support

[unreleased]: https://github.com/trufi-association/arb-convert/compare/v1.1.2...HEAD
[1.1.2]: https://github.com/trufi-association/arb-convert/compare/v1.1.1...v1.1.2
[1.1.1]: https://github.com/trufi-association/arb-convert/compare/v1.1.0...v1.1.1
[1.1.0]: https://github.com/trufi-association/arb-convert/compare/v1.0.1...v1.1.0
[1.0.1]: https://github.com/trufi-association/arb-convert/compare/v1.0.0...v1.0.1
[1.0.0]: https://github.com/trufi-association/arb-convert/releases/tag/v1.0.0


src/cli/arb2xliff.ts
#!/usr/bin/env node

import commander from 'commander';
import fs from 'fs';
import { convertFromArb } from '../index';
import parseLocale from '../util/parseLocale';

const program = new commander.Command();
program
  .name('arb2xliff')
  .option('--sourcefile <filename>', 'source ARB file (required)')
  .option('--targetfile <filename>', 'target ARB file')
  .option('--original <value>', 'where the translations come from')
  .option('--sourcelang <locale>', 'source locale override, e.g. en-US, in case it cannot be determined from file content or file name')
  .option('--targetlang <locale>', 'target locale override, e.g. de-DE, in case it cannot be determined from file content or file name')
  .option('--outversion <version>', 'XLIFF version, e.g. 1.2 (default) or 2.0/2.1')
  .option('--out <filename>', 'write XLIFF to file if given or stdout if omitted')
  .parse(process.argv);

// No params
if (program.rawArgs.length <= 2) {
  program.help(); // shows help and exits
}

try {
  if (!program.sourcefile) {
    throw new Error("option '--sourcefile <filename>' is required");
  }

  const format = `xliff-${program.outversion || '1.2'}`;
  const sourceContent = fs.readFileSync(program.sourcefile, 'utf8');
  const targetContent = program.targetfile && fs.readFileSync(program.targetfile, 'utf8');
  const result = convertFromArb(format, {
    source: sourceContent,
    target: targetContent,
    original: program.original,
    sourceLanguage: parseLocale(
      program.sourcelang,
      determineArbLocale(sourceContent),
      program.sourcefile,
    ),
    targetLanguage: program.targetfile && parseLocale(
      program.targetlang,
      determineArbLocale(targetContent),
      program.targetfile,
    ),
  });

  if (program.out) {
    fs.writeFileSync(program.out, result.content);
  } else {
    process.stdout.write(result.content);
  }
} catch (error) {
  process.stdout.write(`error: ${error.message}`);
  process.exit(1);
}

function determineArbLocale(content: string) {
  const matches = content.match(/"@@locale":\s*"(.+)"/);

  if (matches) {
    return matches[1];
  }

  return '';
}
